akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -388,4 +405,60 @@ private[akka] class PersistentShard(
     }
   }
 
+  private def retrieveEntityRecoveryConfig(): Config = {
+    val path = settings.tuningParameters.entityRecoveryStrategyConfigPath
+    if (path.isEmpty)
+      ConfigFactory.empty()
+    else
+      context.system.settings.config.getConfig(path)
+  }
+}
+
+abstract class EntityRecoveryStrategy {
+  import ShardRegion.EntityId
+  import akka.actor.ActorContext
+  def recoverEntities(context: ActorContext, entities: Set[EntityId])
```

Comment:
I think we can still implement #15797 if we keep the APIs internally extensible, but not "totally extensible" as Patrik suggested.

I also agree with the comment about ActorContext and the API returning a Future.
// Patrik is on vacation for a while now, I'll take over reviewing here :)